### ISSUE TITLE
 Force a trip if another is done - Python - Ubuntu - Routing

### ISSUE THREAD

**Issue by Skyloq** [google/or-tools#2007]
Hello,

I use python

Here is my problem: I would like to have a vehicle transporting an object from point A to point B and then from point B to point A.

I already have this : 

![image](https://user-images.githubusercontent.com/39385087/81275734-5d953900-9052-11ea-9960-43936745c516.png)

and this : 

![image](https://user-images.githubusercontent.com/39385087/81275885-91705e80-9052-11ea-8acd-69fffb9c0370.png)

Which means that my trips are optional.

Take 4 point: 1,2,3,4

We have 1-> 2 and 3-> 4, so : 

`routing.AddPickupAndDelivery(1, 2)`

`routing.AddPickupAndDelivery(3, 4)`

I would like these trips to stay optionnal but impose the following constraint on the solver: If one trip then the other

In other words, ([1-> 2] && [3-> 4]) or Nothing

I have several tracks: modify the `AddDisjunction ()` or add an `AddPickupAndDelivery (2,4)` which has not worked so far

Thank you for reading and sorry for my poor English

**Comment by Mizux** [comment]
did you try to add that ?
```python
routing.solver().Add(
  routing.VehicleVar(manager.NodeToIndex(2) == routing.VehicleVar(manager.NodeToIndex(3))
```
in order to "chain" `2` and `3`

Otherwise you should ask on the mailing-list !

**Comment by Skyloq** [comment]
Thank you for answering

I tried something similar: `routing.solver().Add(routing.VehicleVar(2)
                             == routing.VehicleVar(4))`

I will try your code right away

**Comment by Mizux** [comment]
To drop all the circuit or note you can use Disjunction with cardinality of 4
```python
routing.AddDisjunction([A,B,C,D], 10000, 4)
```
see:
https://github.com/google/or-tools/blob/5ff76b487a6c2006326765d6417964599eedc8c9/ortools/constraint_solver/routing.h#L570-L588

**Comment by Skyloq** [comment]
I will try this too, thanks again

**Comment by Mizux** [comment]
also since node are added one by one you should consider checking if node are active to apply the constraint
see: https://github.com/google/or-tools/issues/847#issuecomment-420608114

**Comment by Skyloq** [comment]
I have tried these different methods, but none of them work yet. Which is probably due to an error coming from me.

I did not find you on Discord. Here is my profile: Skyloq#8629

**Comment by lperron** [comment]
This whole discussion is lost for the community. Please use the mailing list.

**Comment by parcelero** [comment]
> did you try to add that ?
> 
> ```python
> routing.solver().Add(
>   routing.VehicleVar(manager.NodeToIndex(2) == routing.VehicleVar(manager.NodeToIndex(3))
> ```
> 
> in order to "chain" `2` and `3`
> 
> Otherwise you should ask on the mailing-list !

Hey,

im have used this constraint and works good but i also need to add something else to my problem, i havent found a great solution for ti, ill be very thanked if you could give me a hand with id.

i need also to make some nodes/visits to be subsequent not only in the same vehicle.
for example:
bundled_orders = [1,2,3,4], [5,6] [7,8,9][10]

route_1 = 0 -> 1 -> 2 -> 3 -> 4 -> 10
route_2 = 0 -> 9 -> 7 -> 8 -> 5 -> 6

thank you!

**Comment by parcelero** [comment]
> did you try to add that ?
> 
> ```python
> routing.solver().Add(
>   routing.VehicleVar(manager.NodeToIndex(2) == routing.VehicleVar(manager.NodeToIndex(3))
> ```
> 
> in order to "chain" `2` and `3`
> 
> Otherwise you should ask on the mailing-list !

Hey,
hope i find you well!

i have used this constraint and works good but i also need to add something else to my problem, i havent found a great solution for ti, ill be very thanked if you could give me a hand with id.

i need also to make some nodes/visits to be subsequent not only in the same vehicle.
for example:
bundled_orders = [1,2,3,4], [5,6] [7,8,9][10]

route_1 = 0 -> 1 -> 3 -> 2 -> 4 -> 10
route_2 = 0 -> 9 -> 7 -> 8 -> 5 -> 6

thank you!

**Comment by Mizux** [comment]
Did you try to add so precedence constraint: i.e. time at 1 < time at 2 < time at 3 < time at 4 etc... ?
```py
solver = routing.solver()
one = manager.NodeToIndex(1)
two = manager.NodeToIndex(2)
three = manager.NodeToIndex(3)
four = manager.NodeToIndex(4)
solver.Add(time_dim.CumulVar(one) <= time_dim.CumulCar(two))
solver.Add(time_dim.CumulVar(two) <= time_dim.CumulCar(three))
solver.Add(time_dim.CumulVar(three) <= time_dim.CumulCar(four))
...
```

note: if some node can be dropped you may need to "multiply" by `routing.ActiveVar(location_index)` to nullify the expression thus constraint become `0 <= 0` and is still valid...
ref: https://github.com/google/or-tools/blob/405b1dcaf1db6a1165e48a836e931828fe1d6fd5/ortools/constraint_solver/routing.h#L1359-L1360

**Comment by parcelero** [comment]
> Did you try to add so precedence constraint: i.e. time at 1 < time at 2 < time at 3 < time at 4 etc... ?
> 
> ```python
> solver = routing.solver()
> one = manager.NodeToIndex(1)
> two = manager.NodeToIndex(2)
> three = manager.NodeToIndex(3)
> four = manager.NodeToIndex(4)
> solver.Add(time_dim.CumulVar(one) <= time_dim.CumulCar(two))
> solver.Add(time_dim.CumulVar(two) <= time_dim.CumulCar(three))
> solver.Add(time_dim.CumulVar(three) <= time_dim.CumulCar(four))
> ...
> ```
> 
> note: if some node can be dropped you may need to "multiply" by `routing.ActiveVar(location_index)` to nullify the expression thus constraint become `0 <= 0` and is still valid... ref:
> 
> https://github.com/google/or-tools/blob/405b1dcaf1db6a1165e48a836e931828fe1d6fd5/ortools/constraint_solver/routing.h#L1359-L1360

i have tried it but i want it to be able to decide the order.
using this constraints the only way i can route the grouo 1 - 2 - 3 - 4 is in that order and i am trying to make it able to be (for example) 1-4-2-3 
do you know if its possible?
